examples/multigpu_advtrain/model.py

```diff
@@ -0,0 +1,339 @@
+                            initializer, trainable=self.training)
+        return v
+
+    def set_input_shape_ngpu(self, new_input_shape, **kwargs):
+        assert self.device_name
+        device_name = self.device_name
+        if self.input_shape is None:
+            # First time setting the input shape
+            self.input_shape = [None] + [int(d) for d in list(new_input_shape)]
+
+        if device_name in self.params_device:
+            # There is a copy of weights on this device
+            self.__dict__.update(self.params_device[device_name])
+            return
+
+        # stop recursion
+        self.params_device[device_name] = {}
+
+        # Initialize weights on this device
+        with tf.device(device_name):
+            keys_before = self.__dict__.keys()
+            self.set_input_shape(self.input_shape, **kwargs)
+            keys_after = self.__dict__.keys()
+            if self.params_names is None:
+                self.params_names = list(set(keys_after) - set(keys_before))
+            params = dict([(k, self.__dict__[k]) for k in self.params_names])
+            self.params_device[device_name] = params
+
+    def create_sync_ops(self, host_device):
+        sync_ops = []
+        host_params = self.params_device[host_device]
+        for device, params in (self.params_device).iteritems():
+            if device == host_device:
+                continue
+            for k in self.params_names:
+                if isinstance(params[k], tf.Variable):
+                    sync_ops += [tf.assign(params[k], host_params[k])]
+        return sync_ops
+
+    def get_output_shape(self):
+        return self.output_shape
+
+    def fprop(self, x, training=False, **kwargs):
+        self.training = training
+        if self.name is None:
+            self.set_input_shape_ngpu(x.shape[1:], **kwargs)
+            return self.fprop_noscope(x, **kwargs)
+        else:
+            with tf.variable_scope(self.name):
+                self.set_input_shape_ngpu(x.shape[1:], **kwargs)
+                return self.fprop_noscope(x, **kwargs)
+
+
+class Linear(Layer):
+
+    def __init__(self, num_hid, w_name='W', **kwargs):
+        super(Linear, self).__init__(**kwargs)
+        self.num_hid = num_hid
+        self.input_shape = None
+        self.w_name = w_name
+
+    def set_input_shape(self, input_shape, **kwargs):
+        batch_size, dim = input_shape
+        self.input_shape = [batch_size, dim]
+        self.output_shape = [batch_size, self.num_hid]
+        with tf.variable_scope(self.__class__.__name__):
+            shape = [dim, self.num_hid]
+            init = tf.truncated_normal(shape, stddev=0.1)
+            self.W = self.get_variable(self.w_name, init)
+            self.b = self.get_variable('b', .1 + np.zeros(
+                (self.num_hid,)).astype('float32'))
+
+    def fprop_noscope(self, x, **kwargs):
+        return tf.matmul(x, self.W) + self.b
+
+
+class Conv2D(Layer):
```

Comment:
similarly, how hard would it be to use the one from the tutorials?

Reply:
Fixed in 714afc0. Changed to Conv2DnGPU as it has to use the LayernGPUs get_variable.